components/places/src/places.udl

```diff
@@ -145,63 +171,48 @@ message BookmarkNode {
      * - Not provided if `child_nodes` is provided, to avoid sending more data
      *   over the FFI than necessary.
      */
-    repeated string child_guids = 9;
+    sequence<string> child_guids;
 
     /**
      * Data about folder children, in order. Only present for type =
      * `BookmarkType::Folder`.
      *
      * For performance reasons, this only is provided if it's requested.
      */
-    repeated BookmarkNode child_nodes = 10;
-
-    /**
-     * Returned by reads, and used to distinguish between the cases of
-     * "empty child_nodes because the API doesn't return children" and
-     * "empty child_nodes because this folder has no children (but
-     * we'd populate them if it had them)".
-     *
-     * Only required because you can't have `optional repeated`.
-     *
-     * Leaving this out is equivalent to false.
-     */
-    optional bool have_child_nodes = 11;
-}
-
-/** An array of bookmark nodes, since we can't represent that directly */
-message BookmarkNodeList {
-    repeated BookmarkNode nodes = 1;
-}
+    sequence<BookmarkNode>? child_nodes;
+};
 
-// Protobuf allows nesting these, but prost behaves weirdly if we do.
-enum SearchResultReason {
+[Enum]
+interface MatchReason {
     // Never used in practice. Maybe remove this from here and from the rust enum?
-    KEYWORD = 1;
-    ORIGIN = 2;
-    URL = 3;
-    PREVIOUS_USE = 4;
-    BOOKMARK = 5;
-    // If we get real tag support, just add `optional string tags` to SearchResult below, but
+    Keyword();
+    Origin();
+    Url();
+    PreviousUse();
+    Bookmark();
+    // If we get real tag support, just add `string? tags` to SearchResult below, but
     // for now expose that it was because of tags.
-    TAG = 6;
-}
-
-message SearchResultMessage {
-    required string url = 1;
-    required string title = 2;
-    required int64 frecency = 3;
-    repeated SearchResultReason reasons = 4 [packed = true];
-}
-
-message SearchResultList {
-    repeated SearchResultMessage results = 1;
-}
-
-message TopFrecentSiteInfo {
-    required string url = 1;
-    optional string title = 2;
-}
-
-message TopFrecentSiteInfos {
-    repeated TopFrecentSiteInfo infos = 1;
-}
+    Tags(string tags);
+};
+
+dictionary SearchResult {
+    string search_string;
+    string url;
+    string title;
+    string? icon_url;
+    i64 frecency;
+    sequence<MatchReason> reasons;
+};
+
+
+dictionary TopFrecentSiteInfo {
+    string url;
+    string? title;
+};
+
+// This just makes sure we have some extra helper functions available,
+// to be called from our hand-written bindings.
```

Comment:
Now it's UniFFI's turn for some weirdness workarounds!

UniFFI tries not to generate helper code for types that your API doesn't use. We have some functions that want to return e.g. a `sequence<SearchResult>`, so we want UniFFI's helper code for lifting that type. But UniFFI won't generate that helper code unless it sees the type somewhere in the interface. So, we fake it here, at least until the type appears organically in the signature of a uniffied function.